src/Analyser/MutatingScope.php

```diff
@@ -3520,20 +3520,39 @@ private function specifyExpressionType(Expr $expr, Type $type, ?Type $nativeType
 			if ($dimType instanceof ConstantIntegerType || $dimType instanceof ConstantStringType) {
 				$exprVarType = $this->getType($expr->var);
 				if (!$exprVarType instanceof MixedType) {
-					$types = [
-						new ArrayType(new MixedType(), new MixedType()),
-						new ObjectType(ArrayAccess::class),
-					];
-					if ($dimType instanceof ConstantIntegerType) {
-						$types[] = new StringType();
+					if ($exprVarType instanceof IntersectionType && count($exprVarType->getTypes()) > 8) {
```

Comment:
I am not sure this is the fix ondrej imagined, but lets see.

I am adding the old codepath back in here for intersection with more then 8 elements.
this seems to be good enough for the whole test-suite to pass (no regressions) and we get back the performance of bug 5081, we had in 1.8.2